File: benchmarks/beginner/B2-file-line-counter/javascript/main.js
const fs = require("fs");
const path = require("path");

const DATA_DIR = path.join(__dirname, "..", "..", "..", "..", "data");
const FILE_PATH = path.join(DATA_DIR, "b2_lines.txt");
const LINES = 1_000_000;
const RUNS = 10;

function writeFile() {
  fs.mkdirSync(DATA_DIR, { recursive: true });
  const fd = fs.openSync(FILE_PATH, "w");
  for (let i = 1; i <= LINES; i++) {
    fs.writeSync(fd, `Line ${i}: The quick brown fox jumps over the lazy dog\n`);
  }
  fs.closeSync(fd);
}

function readAndCount() {
  const data = fs.readFileSync(FILE_PATH, "utf8");
  let count = 0;
  let idx = 0;
  while (idx < data.length) {
    const next = data.indexOf("\n", idx);
    if (next === -1) {
      if (idx < data.length) count++;
      break;
    }
    count++;
    idx = next + 1;
  }
  return count;
}

function main() {
  const writeTimings = [];
  const readTimings = [];

  console.log("=== BENCHMARK: B2 - File Line Counter ===");
  console.log("Language: JavaScript");
  console.log(`Runtime:  Node ${process.version}`);
  console.log();

  let count = 0;
  for (let i = 1; i <= RUNS; i++) {
    let start = process.hrtime.bigint();
    writeFile();
    const writeElapsed = Number(process.hrtime.bigint() - start) / 1e6;

    start = process.hrtime.bigint();
    count = readAndCount();
    const readElapsed = Number(process.hrtime.bigint() - start) / 1e6;

    writeTimings.push(writeElapsed);
    readTimings.push(readElapsed);
    const total = writeElapsed + readElapsed;
    console.log(
      `Run ${String(i).padStart(2)}: Write ${writeElapsed.toFixed(2)} ms | Read ${readElapsed.toFixed(2)} ms | Total ${total.toFixed(2)} ms`
    );
  }

  console.log();
  for (const [label, timings] of [["Write", writeTimings], ["Read", readTimings]]) {
    const min = Math.min(...timings);
    const avg = timings.reduce((a, b) => a + b, 0) / timings.length;
    const max = Math.max(...timings);
    const variance = timings.reduce((a, t) => a + (t - avg) ** 2, 0) / timings.length;
    const stddev = Math.sqrt(variance);
    console.log(`${label} — Min: ${min.toFixed(2)} ms | Avg: ${avg.toFixed(2)} ms | Max: ${max.toFixed(2)} ms | StdDev: ${stddev.toFixed(2)} ms`);
  }

  const totals = writeTimings.map((w, i) => w + readTimings[i]);
  const min = Math.min(...totals);
  const avg = totals.reduce((a, b) => a + b, 0) / totals.length;
  const max = Math.max(...totals);
  const variance = totals.reduce((a, t) => a + (t - avg) ** 2, 0) / totals.length;
  const stddev = Math.sqrt(variance);
  console.log(`Total — Min: ${min.toFixed(2)} ms | Avg: ${avg.toFixed(2)} ms | Max: ${max.toFixed(2)} ms | StdDev: ${stddev.toFixed(2)} ms`);

  console.log();
  const status = count === LINES ? "✓" : "✗";
  console.log(`Verification: ${count} ${status}`);

  fs.unlinkSync(FILE_PATH);
}

main();

File: benchmarks/beginner/B2-file-line-counter/rust/main.rs
use std::fs;
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::path::PathBuf;
use std::time::Instant;

const LINES: usize = 1_000_000;
const RUNS: usize = 10;

fn data_path() -> PathBuf {
    let mut p = std::env::current_dir().unwrap();
    for _ in 0..4 {
        p.pop();
    }
    p.push("data");
    p.push("b2_lines.txt");
    p
}

fn write_file(path: &PathBuf) {
    fs::create_dir_all(path.parent().unwrap()).unwrap();
    let f = fs::File::create(path).unwrap();
    let mut w = BufWriter::new(f);
    for i in 1..=LINES {
        writeln!(w, "Line {}: The quick brown fox jumps over the lazy dog", i).unwrap();
    }
    w.flush().unwrap();
}

fn read_and_count(path: &PathBuf) -> usize {
    let f = fs::File::open(path).unwrap();
    let reader = BufReader::new(f);
    reader.lines().count()
}

fn main() {
    let path = data_path();
    let mut write_timings = [0.0f64; RUNS];
    let mut read_timings = [0.0f64; RUNS];

    println!("=== BENCHMARK: B2 - File Line Counter ===");
    println!("Language: Rust");
    println!("Runtime:  rustc (native)");
    println!();

    let mut count = 0usize;
    for i in 0..RUNS {
        let start = Instant::now();
        write_file(&path);
        let write_elapsed = start.elapsed().as_secs_f64() * 1000.0;

        let start = Instant::now();
        count = read_and_count(&path);
        let read_elapsed = start.elapsed().as_secs_f64() * 1000.0;

        write_timings[i] = write_elapsed;
        read_timings[i] = read_elapsed;
        let total = write_elapsed + read_elapsed;
        println!("Run {:2}: Write {:.2} ms | Read {:.2} ms | Total {:.2} ms", i + 1, write_elapsed, read_elapsed, total);
    }

    println!();
    for (label, timings) in [("Write", &write_timings), ("Read", &read_timings)] {
        let min = timings.iter().cloned().fold(f64::INFINITY, f64::min);
        let max = timings.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        let avg = timings.iter().sum::<f64>() / RUNS as f64;
        let variance = timings.iter().map(|t| (t - avg).powi(2)).sum::<f64>() / RUNS as f64;
        let stddev = variance.sqrt();
        println!("{} — Min: {:.2} ms | Avg: {:.2} ms | Max: {:.2} ms | StdDev: {:.2} ms", label, min, avg, max, stddev);
    }

    let totals: Vec<f64> = (0..RUNS).map(|i| write_timings[i] + read_timings[i]).collect();
    let min = totals.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = totals.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let avg = totals.iter().sum::<f64>() / RUNS as f64;
    let variance = totals.iter().map(|t| (t - avg).powi(2)).sum::<f64>() / RUNS as f64;
    let stddev = variance.sqrt();
    println!("Total — Min: {:.2} ms | Avg: {:.2} ms | Max: {:.2} ms | StdDev: {:.2} ms", min, avg, max, stddev);

    println!();
    let status = if count == LINES { "✓" } else { "✗" };
    println!("Verification: {} {}", count, status);

    fs::remove_file(&path).ok();
}
